### PR TITLE
feat(lsp): add fswatch watchfunc backend

### PR DESCRIPTION
**Pull request by lewis6991**
### Problem

`vim._watch.poll` has terrible performance.

### Solution

On linux use fswatch as a watcher backend if available.

---

- Add File watcher section to health:vim.lsp. Warn if watchfunc is watchdirs.
- Renamed `_watch.poll` -> `_watch.watchdirs`
- All watchfuncs now support `include_pattern` and `exclude_pattern`.
- Improve type annotations all around.
- Run related tests on all watchfunc backends
- Tidied and simplified tests
- Close the file watchers when the LSP client is stopped.

Relates to: #23291

### Future work

Add a [watchmen](https://facebook.github.io/watchman/) [watchfunc](https://github.com/neovim/neovim/issues/23291#issuecomment-1712422887)

Note to reviewers: Please review by commit.

### CODE REVIEW

File: runtime/lua/vim/lsp/health.lua
Comment (mfussenegger) [inline runtime/lua/vim/lsp/health.lua:60]: I think we should rename the method, it's not actually polling anymore but opposed to the watch method it sets up a watcher per directory

**Review by mfussenegger — COMMENTED** [review]
Is there any way we can have a test case to ensure this results in the expected triggers?

**Comment by lewis6991** [comment]
I'm happy to update any existing tests so they cover both backends.

File: runtime/lua/vim/_watch.lua
Comment (MariaSolOs) [inline runtime/lua/vim/_watch.lua:11]: I think we should still keep the description.

File: runtime/lua/vim/_watch.lua
Comment (MariaSolOs) [inline runtime/lua/vim/_watch.lua:202]: Isn't this redundant?
```suggestion
---@param opts watch.PollOpts? Additional options
```

File: runtime/lua/vim/_watch.lua
Comment (MariaSolOs) [inline runtime/lua/vim/_watch.lua:227]:
```suggestion
---@param callback function The function called when new events occur.
```

File: runtime/lua/vim/_watch/fswatch.lua
Comment (MariaSolOs) [inline runtime/lua/vim/_watch/fswatch.lua:1]: I would appreciate some comments in this file :)

File: runtime/lua/vim/_watch.lua
Comment (lewis6991) [reply runtime/lua/vim/_watch.lua:202]: No. it's to control the doc-gen. Which this `(table|nil)` will be documented.

Comment (MariaSolOs) [reply runtime/lua/vim/_watch.lua:202]: In that case why keep the `?`?

File: runtime/lua/vim/_watch.lua
Comment (lewis6991) [reply runtime/lua/vim/_watch.lua:11]: I dunno, it seems pretty redundant since know it's an enum, we know it's about types of events and it's defined in `_watch.lua` so must be used by watchers.

File: runtime/lua/vim/_watch.lua
Comment (lewis6991) [reply runtime/lua/vim/_watch.lua:202]: Because that is a part of the type.

Comment (lewis6991) [reply runtime/lua/vim/_watch.lua:202]: Ah, we don't generate docs from this file so it is redundant.

File: runtime/lua/vim/_watch.lua
Comment (MariaSolOs) [reply runtime/lua/vim/_watch.lua:11]: Right but it's part of the "public API" of this module (and yeah the module itself is private but still).

File: runtime/doc/news.txt
Comment (MariaSolOs) [inline runtime/doc/news.txt:339]: Should we maybe also mention here that `fswatch` is recommended?
```suggestion
  • On Linux, `fswatch` (recommended) is used a the backend if available, otherwise
```

File: runtime/lua/vim/_watch.lua
Comment (MariaSolOs) [reply runtime/lua/vim/_watch.lua:202]: > Ah, we don't generate docs from this file so it is redundant.

So does this mean that you agree with me on removing the `(table|nil)`?

**Comment by MariaSolOs** [comment]
Unrelated nit but since we're already touching this file, is this function really needed? It's a simple one-liner used in a single place.

https://github.com/neovim/neovim/blob/57affc3307a0d6b3a60fe4348a8d3fde8d6f2891/runtime/lua/vim/_watch.lua#L14-L20

File: runtime/lua/vim/_watch.lua
Comment (lewis6991) [reply runtime/lua/vim/_watch.lua]: Yes, as like you said it is redundant. I assumed because the docstring was complete that we exported docs for this.

File: runtime/lua/vim/_watch.lua
Comment (lewis6991) [reply runtime/lua/vim/_watch.lua:11]: Done.

File: runtime/lua/vim/_watch.lua
Comment (lewis6991) [inline runtime/lua/vim/_watch.lua:53]: I don't think this should be included and that all watchfuncs should try to support the same set of options. I've begun that here by adding universal support for `include_pattern` and `exclude_pattern`.

**Comment by lewis6991** [comment]
Trying to get all the backends crossed with all the platforms to agree on exact events was a bit tricky since different API calls can cause different events in any one combination.

For example `io.open` on Linux appears to cause a `Create` event with `fswatch`, however Windows with any backend requires `file:close()` to cause the `Created` event.

In practice these "event inconsistencies" are benign but does make testing a little awkward.

I've worked around it by creating files by renaming a temporary file which is how vim saves files internally anyway.

Hopefully that makes CI happy give or take some more random delays. 🤞

File: runtime/lua/vim/_watch.lua
Comment (MariaSolOs) [reply runtime/lua/vim/_watch.lua]: Then why haven't you removed the redundancy?

**Comment by lewis6991** [comment]
Just found we weren't cancelling the file watchers when closing the client. Added a fix here.

**Review by MariaSolOs — APPROVED** [review]
Looking forward to this. I've been following #23291 for a very long time :)

**Comment by lewis6991** [comment]
@mfussenegger I know you still want to explore a LibUV only approach, but until that is implemented can we merge this for the meantime to unblock anyone experiencing the severe performance issues.

If a viable LibUV replacement becomes available, then we can always walk this patch back.

File: runtime/doc/news.txt
Comment (iurimateus) [inline runtime/doc/news.txt:341]: typo?
```suggestion
  • On Linux, `fswatch` (recommended) is used as the backend if available,
```

**Comment by mfussenegger** [comment]
> @mfussenegger I know you still want to explore a LibUV only approach, but until that is implemented can we merge this for the meantime to unblock anyone experiencing the severe performance issues?
> 
> If a viable LibUV replacement becomes available, then we can always walk this patch back.

Yes fine to go ahead with this. Sorry, I didn't mean to imply that this should wait or not be done.
I think even if we have a good-enough libUV story for linux built-in, fswatch is probably still valuable to have as it allows to share watchers across processes (I think?)

I did want to give this a closer look, but on a glance nothing stood out so fine to merge. We can always follow up if something pops up.

Nice choice on the `watchdirs` naming. I wasn't sure what to call it instead.

**Review by mfussenegger — APPROVED** [review]

**Review by justinmk — APPROVED** [review]
This is awesome and hopefully opens the door for exposing a [public file-watcher](https://github.com/neovim/neovim/issues/1380#issuecomment-546667155) interface (also for 'autoread').

Do I understand correctly that libuv seems to work well on mac/win, so we only need a fallback on Linux? Thus we wouldn't need to expose the "provider" concept (unlike e.g. clipboard), though we _internally_ have a fallback.

Does this address any of these items?: https://github.com/neovim/neovim/issues/23291#issuecomment-1817873088

> * De-duplicate debounce (`_watchfiles` has it's own mechanism)
> * Global file-path based watcher de-duplication. ...
> * Deal with watcher limits (`proc/sys/fs/inotify/max_user_watches`)
> * ✅  Configuration for excludes? ...

File: runtime/lua/vim/_watch.lua
Comment (lewis6991) [reply runtime/lua/vim/_watch.lua:26]: This is all private for now, but yes, I think this area needs more work.

Preferably, the LPEG thing should be an implementation detail and these fields should just be `string|string[]`.

**Comment by mfussenegger** [comment]
>> Deal with watcher limits (proc/sys/fs/inotify/max_user_watches)
> 
> Nope, I believe this is a fundamental limit we can't work around. Again, this might improve with fanotify, but is mostly out of our control.

That bullet point was more about ensuring that the error is handled in some sensible way - e.g. aborting instead of stubbornly continue trying to setup additional handles and giving the user a sensible message - telling them to increase the limit, or inform about the limitation. I'm not sure how it behaves right now.

**Comment by asmodeus812** [comment]
By the way @lewis6991 worth nothing that the fs_event performance is abysmal on windows, compared to the fs_poll, however the poll i find is very often incorrect and does not report paths which are supposedly watched in comparison with fs_event. There is also this which i hope is going to be fixing the performance issues with fs_watch on windows - https://github.com/libuv/libuv/pull/4327

**Comment by lewis6991** [comment]
I think `fswatch` uses the same method as LibUV on Windows, so I doubt you'll see any improvements.

**Pull request merged**
